ballista/rust/scheduler/src/flight_sql.rs

```diff
@@ -0,0 +1,460 @@
+        Self {
+            server,
+            _statements: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl FlightSqlService for FlightSqlServiceImpl {
+    type FlightService = FlightSqlServiceImpl;
+    // get_flight_info
+    async fn get_flight_info_statement(
+        &self,
+        query: CommandStatementQuery,
+        _request: FlightDescriptor,
+    ) -> Result<Response<FlightInfo>, Status> {
+        debug!("Got query:\n{}", query.query);
+
+        // Run query
+        let config_builder = BallistaConfig::builder();
+        let config = config_builder
+            .build()
+            .map_err(|e| Status::internal(format!("Error building config: {}", e)))?;
+        let ctx = self
+            .server
+            .state
+            .session_manager
+            .create_session(&config)
+            .await
+            .map_err(|e| {
+                Status::internal(format!("Failed to create SessionContext: {:?}", e))
+            })?;
+        let plan = ctx
+            .sql(query.query.as_str())
+            .await
+            .and_then(|df| df.to_logical_plan())
+            .map_err(|e| Status::internal(format!("Error building plan: {}", e)))?;
+
+        // enqueue job
+        let job_id = self.server.state.task_manager.generate_job_id();
+
+        self.server
+            .state
+            .task_manager
+            .queue_job(&job_id)
+            .await
+            .map_err(|e| {
+                let msg = format!("Failed to queue job {}: {:?}", job_id, e);
+                error!("{}", msg);
+
+                Status::internal(msg)
+            })?;
+
+        let query_stage_event_sender = self
+            .server
+            .query_stage_event_loop
+            .get_sender()
+            .map_err(|e| {
+                Status::internal(format!(
+                    "Could not get query stage event sender due to: {}",
+                    e
+                ))
+            })?;
+
+        query_stage_event_sender
+            .post_event(QueryStageSchedulerEvent::JobQueued {
+                job_id: job_id.clone(),
+                session_id: ctx.session_id().clone(),
+                session_ctx: ctx,
+                plan: Box::new(plan.clone()),
+            })
+            .await
+            .map_err(|e| {
+                let msg =
+                    format!("Failed to send JobQueued event for {}: {:?}", job_id, e);
+                error!("{}", msg);
+                Status::internal(msg)
+            })?;
+
+        // let handle = Uuid::new_v4();
+        // let mut statements = self.statements.try_lock()
+        //     .map_err(|e| Status::internal(format!("Error locking statements: {}", e)))?;
+        // statements.insert(handle, plan.clone());
+
+        // poll for job completion
+        let mut num_rows = 0;
+        let mut num_bytes = 0;
+        let fieps = loop {
+            sleep(Duration::from_millis(100)).await;
```

Comment:
The busy wait is a bit cheesy, but it appears Arrow FlightSQL expects synchronous execution? At least we're in a coroutine context and this isn't blocking a thread, thanks to Tokio, but the polling has been moved out of the client and into the scheduler. I'm not sure how else to do it, just something to be aware of.

Reply:
In this case we may be able to just use a `watch` and let the state backend push status updates (instead of polling). You can wrap that in a `JobCompletionFuture`. Something like 

```
struct JobCompletionFuture {
    job_completion_watch: Watch,
}

impl Future for JobCompletionFuture {
    type Output = Result<JobStatus, BallistaError>;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
            match self.job_completion_watch.poll_next_unpin(cx) {
                Poll::Ready(Some(WatchEvent::Put(_key, value))) => {
                   if let Ok(job_status) = decode_protobuf::<JobStatus>(&value) {
                        Poll::Ready(Ok(job_status))
                    } else {
                       Poll::Ready(Err(BallistaError::Internal("some error".to_owned())))
                   }
                }
                Poll::Ready(None) => Poll::Ready(Err(BallistaError::Internal("watch ended with no update".to_owned()))),
                _ => Poll::Pending,
            }
    }
}
```

It would have to be slightly more complicated than this since we would need to watch both the `CompletedJobs/<job id>` and `FailedJob/<job id>` keys but the basic premise would still work I think.